### PR TITLE
Blog post: Comparison lantern vs pgvector

### PR DESCRIPTION
**Pull request by rzavalet**
Hi! This is the PR for the lantern vs pgvector blog post. 

The google docs draft can be found here: https://docs.google.com/document/d/1HTFod9QcGocGgpeyhQpP7144r1WebfdB48pkSEGZzsg/edit?usp=sharing

Additional material:
- Modified benchmark: https://github.com/binidxaba/ann-benchmarks
- Complete results: https://github.com/binidxaba/lantern-pgvector-comparison

Any additional feedback is appreciated!!

### CODE REVIEW

**Comment by samay-sharma** [comment]
@rzavalet : Is it good for me to take a pass to review? Or are there comments from the google doc still left to resolve?

**Comment by rzavalet** [comment]
> @rzavalet : Is it good for me to take a pass to review? Or are there comments from the google doc still left to resolve?

Yes! Please have a look when you get a chance. This PR addresses all feedback in the google doc. 

Let me know if there is anything else that can be improved.

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+![Baseline index size](./002-baseline-index-size.png)
+
+That’s cool… And what about the resulting throughput and recall? Here are the results:
+
+![Baseline recall](./003-baseline-recall.png)
+
+![Baseline throughput](./004-baseline-qps.png)
+
+Ok, the recall is similar, however the resulting QPS for Lantern is worse when compared to Pgvector. Pgvector can process between 72% and 81% more queries per second.
```
Comment (ChuckHend) [inline docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:95]: just a suggestion, but maybe say "Pgvector outperforms Lantern in QPS", or something similar. Feels more positive in tone.

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+And again, Pgvector has better throughput.
+
+![Throughput](./014-qps-m16.png)
+
+Increasing the m parameter to 24, we get similar conclusions:
+
+![Recall](./015-recall-m24.png)
+
+![Throughput](./016-qps-m24.png)
```
Comment (ChuckHend) [inline docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:120]: Could add a note between the recall and throughput graphs, something like "pgvector has higher throughput in all values of ef_search"

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+For a quantitative comparison, I took the [ANN-Benchmarks](https://github.com/erikbern/ann-benchmarks) and [extended it](https://github.com/binidxaba/ann-benchmarks) to support the Lantern extension. It was just a matter of creating a copy of the Pgvector directory, naming it Lantern and making a few adjustments to use the corresponding API.
+
+The benchmark starts a container that has Postgres 14 installed and then it enables the corresponding extension. Then, it inserts a dataset into a table and builds a vector index (in both cases an HNSW index). After that, it executes a bunch of queries and evaluates the recall.
+
+Along the way, the benchmark collects several metrics, such as:
+
+- Build Time
+- Index Size
+- Recall
+- Throughput
```
Comment (ChuckHend) [inline docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:64]: Do you happen to have latency results as well? That would also be interesting to see IMO.

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md [hunk 4]
```diff
@@ -0,0 +1,153 @@
+---
+slug: postgres-vector-search-pgvector-and-lantern
+title: 'Comparing Postgres Vector Search approaches: Pgvector vs Lantern'
```
Comment (ChuckHend) [inline docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:3]:
```suggestion
title: 'Benchmarking Postgres Vector Search approaches: Pgvector vs Lantern'
```

Thoughts on calling out Benchmarking in the title? I'd say this is a benchmarking blog :)

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md (hunk 1 above)
Comment (rzavalet) [reply docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md]: Agreed. I incorporated this change.

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md (hunk 2 above)
Comment (rzavalet) [reply docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:127]: Done!

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md (hunk 3 above)
Comment (rzavalet) [reply docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:64]: The benckmark gets percentiles. I incorporated p95 plots.

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md (hunk 4 above)
Comment (rzavalet) [reply docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md]: I changed it to "Benchmarking". @samay-sharma wdyt?

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md [hunk 5]
```diff
@@ -0,0 +1,162 @@
+---
+slug: postgres-vector-search-pgvector-and-lantern
+title: 'Benchmarking Postgres Vector Search approaches: Pgvector vs Lantern'
+authors: [rjzv]
+tags: [postgres, extensions, vector, lantern, pgvector, hnsw]
+image: ./elephant-with-lantern.jpeg
```
Comment (samay-sharma) [inline docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:6]: Can you work with Abby to have a better on-brand image?

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+---
+
+![An elephant holding a lantern](./elephant-with-lantern.jpeg)
+<p align="right" style={{color: 'grey'}}><em>
+Image credit: Generated with Bing Image Creator
+</em></p>
+
+Today, let's talk about Vector Search in Postgres from yet another perspective. In this blog post, we'll cover:
```
Comment (samay-sharma) [inline docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:14]: We should have a better start to the post.

Vector search in Postgres is a space that has seen very active development in the last few months. While pgvector is known to most people, a few weeks ago we came across Lantern which also builds a Postgres based vector database. So, we thought about benchmarking both to compare the two approaches. In this post, we'll cover:

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+- Quick background on doing vector search in PostgreSQL.
+- Comparing Pgvector and Lantern in terms of their syntax and ease of use.
+- Benchmarks comparing index creation time, size, throughput and recall.
+- Summary of the results
+
+## Vector Search in PostgreSQL
```
Comment (samay-sharma) [inline docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:21]: Intro to Vector Search in PostgreSQL

Comment (samay-sharma) [reply docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:21]: Let's also avoid repetition based on what I've asked to change in the intro.

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+That’s cool… And what about the resulting throughput, latency and recall? Here are the results:
+
+![Baseline recall](./003-baseline-recall.png)
+
+![Baseline throughput](./004-baseline-qps.png)
+
+![Baseline p95](./005-baseline-p95.png)
+
+Ok, the recall is similar, however Pgvector outperforms Lantern in QPS. Specifically, Pgvector can process between 72% and 81% more queries per second.
```
Comment (samay-sharma) [inline docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:98]: Let's also change the text to incorporate latency comparison.

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+![Build time](./011-build-time.png)
+
+![Index Size](./012-index-size.png)
+
+Also, as before, the recall is similar when `ef_search` varies (although Pgvector’s is a bit better). This is the graph for construction parameters `ef_construction=200` and `m=16`. As the `ef_search` is increased, the recall of both indexes get closer.
+
+![Recall](./013-recall-m16.png)
+
+
+And again, Pgvector has better throughput and latencies.
```
Comment (samay-sharma) [inline docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:115]: can we say X% better throughput and Y% better latencies.

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+![Throughput](./016-qps-m24.png)
+
+![Latency](./018-p95-m24.png)
+
+These results were consistent with my observations when using Gist-960 and Glove-200. You can see more results [here](https://github.com/binidxaba/lantern-pgvector-comparison/tree/main).
+
+
+## Conclusions
+
+Pgvector is the most popular Postgres extension for vector search. At the time of this writing, the github repository counts 7.2K stars and is actively being discussed on the web. It is also supported on most managed Postgres providers (including [Tembo Cloud](https://tembo.io/)) so is easier for you to access.
```
Comment (samay-sharma) [inline docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:136]: Replace link with cloud.tembo.io

**Review by samay-sharma — COMMENTED** [review]
Mostly looks good. I primarily had feedback on the intro and a few other minor points. I think it'll be good to go after addressing these.

**Review by ChuckHend — APPROVED** [review]

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md (hunk 6 above)
Comment (rzavalet) [reply docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md]: Done

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md (hunk 8 above)
Comment (rzavalet) [reply docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md]: Done

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md (hunk 9 above)
Comment (rzavalet) [reply docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md]: Done

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md (hunk 10 above)
Comment (rzavalet) [reply docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md]: Done

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md (hunk 7 above)
Comment (rzavalet) [reply docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md]: Done

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+But Pgvector is not the only one in this space. The community has come up with other alternatives as well. One such example was [pg_embedding](https://github.com/neondatabase/pg_embedding). In July, [Neon](https://neon.tech/) published a [post](https://neon.tech/blog/pg-embedding-extension-for-vector-search) where they showed how its extension was 20x faster than Pgvector by using HNSW indexes . However, Pgvector quickly caught up and introduced the HNSW index as well. Around that time, on September 29th, Neon stopped supporting `pg_embedding` and suggested migrating to Pgvector.
+
+Another Postgres extension in the vector search domain is [Lantern](https://github.com/lanterndata/lantern), developed by a [company of the same name](https://lantern.dev/). In October, a post on their blog claimed that their extension could outperform Pgvector by 90x in index creation time. That was an outstanding achievement!
```
Comment (samay-sharma) [inline docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md:29]: If we're referring to the blog post, let's also link to it.

Comment (rzavalet) [reply docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md]: Done!

File: docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+
+
+My takeaway is that if you are familiar with one of the extensions, you can easily understand how to use the other. No major complications are expected on that front.
+
+
+## Leveraging ANN-Benchmarks
+
+For a quantitative comparison, I took the [ANN-Benchmarks](https://github.com/erikbern/ann-benchmarks) and [extended it](https://github.com/binidxaba/ann-benchmarks) to support the Lantern extension. It was just a matter of creating a copy of the Pgvector directory, naming it Lantern and making a few adjustments to use the corresponding API.
+
+The benchmark starts a container that has Postgres 14 installed and then it enables the corresponding extension. Then, it inserts a dataset into a table and builds a vector index (in both cases an HNSW index). After that, it executes a bunch of queries and evaluates the recall.
```
Comment (samay-sharma) [inline docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md:57]: It says Postgres 15 at the bottom and Postgres 14 here. Which one is correct?

File: docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+![An elephant holding a lantern](./elephant-with-lantern.jpeg)
+<p align="right" style={{color: 'grey'}}><em>
+Image credit: Generated with Bing Image Creator
+</em></p>
+
+Vector search in Postgres is a space that has seen very active development in the last few months. While Pgvector is known to most people, a few weeks ago we came across Lantern, which also builds a Postgres-based vector database. So, we thought about benchmarking both to compare the two approaches. In this post, we'll cover:
+
+- Quick background on doing vector search in PostgreSQL.
+- Comparing Pgvector and Lantern in terms of syntax and ease of use.
+- Benchmarks comparing index creation time, size, throughput, and recall.
```
Comment (samay-sharma) [inline docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md:18]: Let's put latency here too.

File: docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+- Summary of the results
+
+## Intro to Vector Search in PostgreSQL
```
Comment (samay-sharma) [inline docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md:21]: Let's see if there are any opportunities to link to your vector related blog posts in this section. Let's not write about them but just find opportunities to link to them.

File: docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+And once again, Pgvector has 42-58% higher throughput and 30-39% better latencies for all values of `ef_search`:
+
+![Throughput](./017-sift-qps-m24.png)
+![Latency](./017-sift-latency-m24.png)
+
+The general trends were consistent with my observations when using Gist-960 and Glove-200. You can see more results [here](https://github.com/binidxaba/lantern-pgvector-comparison/tree/main).
+
+
+## Conclusions
+
```
Comment (samay-sharma) [inline docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md:129]: Can we add a table summary of results here which makes it easy to grasp the results of the blog? It's just that we share so many data points in this section that I forget what the numbers were like.

File: docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+Pgvector is the most popular Postgres extension for vector search. At the time of this writing, the github repository counts 7.6K stars and is actively being discussed on the web. It is also supported on most managed Postgres providers (including [Tembo Cloud](https://cloud.tembo.io/)) so is easier for you to access.
+
+Lantern is a young project that leverages the popular [USearch engine](https://github.com/unum-cloud/usearch/). As of today, the extension has been starred more than 400 times in github and is in very active development. 
+
+Both extensions offer similar API and support the HNSW index. According to my experiments (Pgvector 0.5.1 and Lantern 0.0.11), Lantern's index creation is faster and produces smaller indexes. However, Pgvector provides better recall and throughput.
```
Comment (samay-sharma) [inline docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md:134]: Comment on latency too.

**Review by samay-sharma — COMMENTED** [review]
What I believe should be the last round of feedback. Once you have the image and have addressed these, I'll merge.

File: docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md (hunk 12 above)
Comment (binidxaba) [reply docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md]: It is postgres 15. Fixed.

File: docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md (hunk 13 above)
Comment (binidxaba) [reply docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md]: Done

File: docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md (hunk 14 above)
Comment (binidxaba) [reply docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md:21]: Added a couple of links.

File: docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md (hunk 15 above)
Comment (binidxaba) [reply docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md:129]: Done, let me know what you think.

File: docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md (hunk 16 above)
Comment (binidxaba) [reply docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md]: Done

File: docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md (hunk 15 above)
Comment (rzavalet) [reply docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md:129]: (Sorry, I was using another account for my replies)

File: docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md (hunk 5 above)
Comment (rzavalet) [reply docusaurus/blog/2024-01-11-pgvector-and-lantern/index.md]: Done!

**Review by samay-sharma — APPROVED** [review]

File: docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md (hunk 15 above)
Comment (samay-sharma) [reply docusaurus/blog/2024-01-18-pgvector-and-lantern/index.md]: looks good

**Review by samay-sharma — APPROVED** [review]

**Pull request merged**